inference-engine/src/legacy_api/include/graph_tools.hpp

```diff
@@ -128,30 +128,45 @@ inline OutInfoWrapper default_order(CNNLayer* layer) {
 template <class T, class Ordering = std::function<OutInfoWrapper(CNNLayer*)>>
 inline bool DFS(std::unordered_map<CNNLayer*, bool>& visited, const InferenceEngine::CNNLayerPtr& layer, const T& visit,
                 bool visitBefore, const Ordering& order = &default_order) {
+
     if (layer == nullptr) {
         return true;
     }
 
-    if (visitBefore) visit(layer);
-    visited[layer.get()] = false;
-    for (auto outLayerPtr : order(layer.get())) {
-        auto i = visited.find(outLayerPtr.get());
-        if (i != visited.end()) {
-            /**
-             * cycle detected we entered still not completed node
-             */
-            if (!i->second) {
-                return false;
+    auto list = std::list<CNNLayerPtr>{layer};
+    while (!list.empty()) {
+        const auto& currentNode = list.front();
+        const auto& currentLayer = currentNode.get();
+
+        const auto& visitState = visited.find(currentLayer);
```

Comment:
Looks like in this case `visited` can be a local variable and we can remove it from function arguments.

Reply:
Agree, it'd make API more clear. I even would like go further and would try to get rid of `visitBefore` argument..

But not in this PR, it's intended to change implementation only

Reply:
Could create a task for this activity?

Reply:
are you keeping reference to temp iterator? please tech me why this is working at all and not read access violation on line 142?

Reply:
Because const ref extending lifetime of temporary objects)